pkg/controller/secret/controller.go

```diff
@@ -0,0 +1,360 @@
+	}
+
+	return &external{client, applicator}, nil
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Secret)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotSecret)
+	}
+
+	if meta.GetExternalName(cr) == "" {
+		return managed.ExternalObservation{}, nil
+	}
+
+	resp, err := e.getSecret(ctx, cr)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(isNotFound, err), errDescribeFailed)
+	}
+
+	generateSecret(resp).Status.AtProvider.DeepCopyInto(&cr.Status.AtProvider)
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	lateInitialize(cr)
+
+	isUpToDate, err := e.isUpToDate(ctx, cr, resp)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errIsUpToDateFailed)
+	}
+
+	cr.Status.SetConditions(v1.Available())
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        isUpToDate,
+		ResourceLateInitialized: !cmp.Equal(cr.Spec.ForProvider, *currentSpec),
+	}, nil
+}
+
+func (e *external) isUpToDate(ctx context.Context, cr *v1alpha1.Secret, resp *secrets.GetSecretOK) (bool, error) {
+	// Styra does not provide an API to retrieve value of a secret.
+	// In order to track changes between the in-cluster value with the upstream version, the controller generates the
+	// checksum of the secret value and the lastModifiedAt timestamp from styra after an update and stores it next to the
+	// referenced secret.
+	// During Observe both checksum are calculated compared with each other. If they match, the secrets are considered equal.
+
+	secretValue, err := e.getSecretValue(ctx, cr)
+	if err != nil {
+		return false, errors.Wrap(err, errGetSecretValueFailed)
+	}
+
+	specCheckSum, err := e.getChecksum(ctx, cr)
+	if err != nil {
+		return false, errors.Wrap(err, errGetChecksumFailed)
+	}
+
+	currentChecksum, err := generateSecretChecksum(secretValue, time.Time(resp.Payload.Result.Metadata.LastModifiedAt))
+	if err != nil {
+		return false, errors.Wrap(err, errGenerateChecksumFailed)
+	}
+
+	return currentChecksum == specCheckSum, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Secret)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotSecret)
+	}
+
+	return managed.ExternalCreation{}, errors.Wrap(e.updateSecret(ctx, cr), errCreateFailed)
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.Secret)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotSecret)
+	}
+
+	return managed.ExternalUpdate{}, errors.Wrap(e.updateSecret(ctx, cr), errUpdateFailed)
+}
+
+func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.Secret)
+	if !ok {
+		return errors.New(errNotSecret)
+	}
+
+	if err := e.deleteChecksum(ctx, cr); err != nil {
```

Comment:
This is fine because we are ignoring the `IsNotFound` error in `deleteChecksum`:

```golang
     return errors.Wrap(client.IgnoreNotFound(e.kube.Delete(ctx, sc)), errDeleteK8sSecretFailed)
```

Reply:
Ah yes. That's if the referenced value secret has been deleted before. Thanks for spotting this!

I fixed this by adding another for `meta.WasDeleted(cr)` before returning the error.